lib/flight_scheduler/job.rb

```diff
@@ -112,8 +117,72 @@ def remove
       end
     end
 
+    def time_out?
+      return false if [nil, 0].include?(@time_out)
+      Process.clock_gettime(Process::CLOCK_MONOTONIC).to_i > @created_time + @time_out
+    end
+
+    # Must be called after adding to the registry
+    def start_time_out_task
+      return if @time_out.nil?
+      Async do |task|
+        Async.logger.info "Job '#{id}' will start timing out in '#{@time_out}'"
+        while FlightScheduler.app.job_registry.lookup_job(id)
+          if @timed_out_time || time_out?
+            if @timed_out_time
+              first = false
+            elsif File.exists? timed_out_path
+              Async.logger.debug "Resuming time out for job: #{id}"
+              @timed_out_time = File.read(timed_out_path).to_i
+              first = false
+            else
+              Async.logger.error "Job Timed Out: #{id}"
+              @timed_out_time = Process.clock_gettime(Process::CLOCK_MONOTONIC).to_i
+              first = true
+            end
+
+            if first
+              send_signal("TERM")
+              File.write(timed_out_path, @timed_out_time)
+              MessageSender.send(command: 'JOB_TIMED_OUT', job_id: id)
+
+              # Allow fast exiting runners to finalise quickly
+              task.yield
+            elsif (Process.clock_gettime(Process::CLOCK_MONOTONIC).to_i - @timed_out_time) > 90
+              send_signal("KILL")
+              # Ensure slow exiting runners have finished
+              task.sleep 5
```

Comment:
This would be better of as sleeping for the correct amount of time.  We know what the current time is and we know at what time the task needs to be timed out.  So this should sleep until that time, which could be less than 5 seconds.  This doesn't need addressing now, but should be addressed at some point.